src/Equinox.Cosmos/Cosmos.fs

```diff
@@ -277,53 +292,157 @@ module Log =
+
+            // Recursively queries for a document by id w/ support for continuation tokens.
+            // Calls tryUpdate(document) as soon as the query returns a document.
+            function tryQueryAndUpdate(continuation) {
+                var query = {query: "select * from root r where r.id = @id and r.p = @p", parameters: [{name: "@id", value: "-1"},{name: "@p", value: batch.p}]};
+                var requestOptions = {continuation: continuation};
+
+                var isAccepted = collection.queryDocuments(collectionLink, query, requestOptions, function (err, documents, responseOptions) {
+                    if (err) throw new Error("Error" + err.message);
+
+                    if (documents.length > 0) {
+                        // If the document is found, update it.
+                        // There is no need to check for a continuation token since we are querying for a single document.
+                        tryUpdate(documents[0], false);
+                    } else if (responseOptions.continuation) {
+                        // Else if the query came back empty, but with a continuation token; repeat the query w/ the token.
+                        // It is highly unlikely for this to happen when performing a query by id; but is included to serve as an example for larger queries.
+                        tryQueryAndUpdate(responseOptions.continuation);
+                    } else {
+                        // Else the snapshot does not exist; create snapshot
+                        var doc = {p:batch.p, id:"-1", i:-1, e:[], c:[]};
+                        tryUpdate(doc, true);
+                    }
+                });
+
+                // If we hit execution bounds - throw an exception.
+                // This is highly unlikely given that this is a query by id; but is included to serve as an example for larger queries.
+                if (!isAccepted) {
+                    throw new Error("The stored procedure timed out.");
+                }
+            }
+
+            function insertEvents(events) {
+                for (i=0; i<events.length; i++) {
+                    var isAccepted = collection.createDocument(collectionLink, events[i], function (err, documentCreated) {
+                        if (err) throw new Error ("Create doc " + JSON.stringify(events[i]) + " failed");
+                    });
+
+                    // If we hit execution bounds - throw an exception.
+                    if (!isAccepted) {
+                        throw new Error("Unable to create document.");
+                    }
+                }
+            }
+
+            // The kernel function
+            function tryUpdate(doc, isCreate) {
+
+                // DocumentDB supports optimistic concurrency control via HTTP ETag.
+                var requestOptions = {etag: doc._etag};
+
+                // Step 1: Update snapshot document's events
+                Array.prototype.push.apply(doc.e, batch.e);
```

Comment:
will this add N events to the tail, or add an array to the tail ? needs to be former

Reply:
Verified this is def the best solution